### PR TITLE
CORE-9954 Enable endpoint changes during re-registration

### PR DESCRIPTION
**Pull request by YashNabar**
Allows members to change their P2P endpoint and re-register with a network to use the new endpoint.

https://github.com/corda/corda-e2e-tests/pull/191

**Test:**

Screenshot 2023-08-15 at 09 31 48

### CODE REVIEW

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt [hunk 1]
```diff
@@ -798,13 +846,50 @@ class StartRegistrationHandlerTest {
...
+        }
+        whenever(pendingMemberInfo.memberProvidedContext).doReturn(contextWithUpdates)
+        whenever(membershipQueryClient.queryMemberInfo(eq(mgmHoldingIdentity.toCorda()), any()))
+            .doReturn(MembershipQueryResult.Success(listOf(activeMemberInfo)))
+        whenever(membershipQueryClient.queryRegistrationRequest(eq(mgmHoldingIdentity.toCorda()), eq(registrationId)))
+            .doReturn(MembershipQueryResult.Success(createRegistrationRequest(serial = 2L)))
+        val result = handler.invoke(registrationState, Record(testTopic, testTopicKey, startRegistrationCommand))
+        result.assertDeclinedRegistration()
+    }
+
```
Comment (williamvigorr3) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt:889]: Should we test that we can update the endpoint. We probably can use most of the logic in: `allows custom properties to be added during re-registration`

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationServiceTest.kt [hunk 2]
```diff
@@ -1037,19 +1051,126 @@ class DynamicMemberRegistrationServiceTest {
...
+
+            postConfigChangedEvent()
+            registrationService.start()
+
+            val exception = assertThrows<InvalidMembershipRegistrationException> {
+                registrationService.register(registrationResultId, member, newContext.toMap())
+            }
+            assertThat(exception).hasMessageContaining("cannot be added, removed or updated")
+        }
+
```
Comment (williamvigorr3) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationServiceTest.kt:1222]: Again should we unit test the case when the endpoint is updated?

**Review by williamvigorr3 — COMMENTED** [review]
I left two minor comments on the tests. Implementation looks good.

**Comment by nikinagy** [comment]
Can we do non-happy path testing as well? Where member tries to re-register with wrong/unaccessible endpoint?

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt (hunk 1 above)
Comment (YashNabar) [reply components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt:889]: Added in [b9a0933](https://github.com/corda/corda-runtime-os/pull/4461/commits/b9a0933a69c188b43907c21dd3e1817706358876)

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationServiceTest.kt (hunk 2 above)
Comment (YashNabar) [reply components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationServiceTest.kt:1222]: Added in [b9a0933](https://github.com/corda/corda-runtime-os/pull/4461/commits/b9a0933a69c188b43907c21dd3e1817706358876)

**Review by williamvigorr3 — APPROVED** [review]
LGTM (when @nikinagy is happy).

**Review by nikinagy — APPROVED** [review]
LGTM

**Pull request merged**